### PR TITLE
Update `other/replace_image_registry`

### PR DESCRIPTION
**Pull request by zeborg**
Signed-off-by: Abhinav Sinha <abhinav@nirmata.com>

## Related issue
During the `1.6.0` tests, this policy was unintentionally removing the `image:tag` after replacing the `registry`.

## Milestone of this PR
`/milestone 1.6.0`

## What type of PR is this
`/kind bug`

## Proposed Changes
Implemented `foreach` and added rules for `Deployment`.

## Checklist

- [x] I have read the [contributing guidelines](https://github.com/kyverno/kyverno/blob/main/CONTRIBUTING.md).
- [ ] I have added tests that prove my fix is effective or that my feature works.
- [ ] My PR contains new or altered behavior to Kyverno and
- [ ] I have read the [PR documentation guide](https://github.com/kyverno/kyverno/blob/main/.github/pr_documentation.md) and followed the process including adding proof manifests to this PR.

### CODE REVIEW

**Comment by zeborg** [comment]
```
➜  kyv git:(main) ✗ k version --short
Client Version: v1.22.2
Server Version: v1.23.3
```

```
I0127 15:57:36.074762       1 logr.go:249]  "msg"="Kyverno"  "Version"="1.6-dev-101-gb2724811"
I0127 15:57:36.075797       1 logr.go:249]  "msg"="Kyverno"  "BuildHash"="main/b27248111fb2903ac85d0ca7467d096767cd37f5"
I0127 15:57:36.075894       1 logr.go:249]  "msg"="Kyverno"  "BuildTime"="2022-01-27_09:26:20pm"
```

```
➜  kyv git:(main) ✗ go run cmd/cli/kubectl-kyverno/main.go test https://github.com/zeborg/policies/other/replace_image_registry -b update-other-replace_image_registry
Enumerating objects: 3110, done.
Counting objects: 100% (806/806), done.
Compressing objects: 100% (468/468), done.
Total 3110 (delta 427), reused 422 (delta 338), pack-reused 2304

Executing replace-image-registry...
applying 1 policy to 3 resources... 
E0127 21:29:35.120983   75876 numeric.go:195] EngineMutate "msg"="Failed to parse from the string key, value is not float, int nor resource quantity" "error"="Version string empty" "kind"="Pod" "namespace"="default" "policy"="replace-image-registry" "rule"="replace-image-registry-pod-initcontainers" "skippedRules"="myapp-pod1" 

skipped mutate policy replace-image-registry -> resource default/Pod/myapp-pod1E0127 21:29:35.124698   75876 numeric.go:195] EngineMutate "msg"="Failed to parse from the string key, value is not float, int nor resource quantity" "error"="Version string empty" "kind"="Deployment" "namespace"="default" "policy"="replace-image-registry" "rule"="replace-image-registry-depl-initcontainers" "skippedRules"="mydeploy" 
E0127 21:29:35.126825   75876 numeric.go:195] EngineMutate "msg"="Failed to parse from the string key, value is not float, int nor resource quantity" "error"="Version string empty" "kind"="Deployment" "namespace"="default" "policy"="replace-image-registry" "rule"="autogen-replace-image-registry-pod-initcontainers" "skippedRules"="mydeploy" 

skipped mutate policy replace-image-registry -> resource default/Deployment/mydeploy
skipped mutate policy replace-image-registry -> resource default/Deployment/mydeploy
│───│────────────────────────│────────────────────────────────────────────│─────────────────────────────│────────│
│ # │ POLICY                 │ RULE                                       │ RESOURCE                    │ RESULT │
│───│────────────────────────│────────────────────────────────────────────│─────────────────────────────│────────│
│ 1 │ replace-image-registry │ replace-image-registry-pod-containers      │ default/Pod/myapp-pod1      │ Pass   │
│ 2 │ replace-image-registry │ replace-image-registry-pod-initcontainers  │ default/Pod/myapp-pod1      │ Pass   │
│ 3 │ replace-image-registry │ replace-image-registry-depl-containers     │ default/Deployment/mydeploy │ Pass   │
│ 4 │ replace-image-registry │ replace-image-registry-depl-initcontainers │ default/Deployment/mydeploy │ Pass   │
│ 5 │ replace-image-registry │ replace-image-registry-pod-containers      │ default/Pod/myapp-pod2      │ Pass   │
│ 6 │ replace-image-registry │ replace-image-registry-pod-initcontainers  │ default/Pod/myapp-pod2      │ Pass   │
│───│────────────────────────│────────────────────────────────────────────│─────────────────────────────│────────│

 Note : test.yaml file name is deprecated in 1.6.0 release. Please provide test yaml file as kyverno-test.yaml 

Test Summary: 6 tests passed and 0 tests failed
```
Seems to be working for me on latest `kyverno:main` branch, not sure why the build is failing.
Although there are some errors related to `nil` JMESPath resolution, the tests have passed with expected results.

I've crosschecked the code as well, and the failing rules are supposed to be skipped anyways due to absence of `initContainers`.

File: other/replace_image_registry/replace_image_registry.yaml [hunk 1]
```diff
@@ -19,15 +19,73 @@ metadata:
...
+                image: |-    
+                        {{ regex_replace_all('^[^/]+', '{{element.image}}', 'myregistry.corp.com' )}}
+    - name: replace-image-registry-pod-initcontainers
+      match:
+        resources:
+          kinds:
+          - Pod
+      preconditions:
+        all:
+        - key: "{{ request.object.spec.initContainers[] | length(@) }}"
```
[the remark this answers is not in the document]
Comment (zeborg) [reply other/replace_image_registry/replace_image_registry.yaml]: Still failing 😓

**Comment by zeborg** [comment]
> Sure it's testing against this commit?

https://github.com/kyverno/policies/blob/cb668e096b350d59aaa0a72b7fd90834adb990a4/.github/workflows/ci.yml#L14-L29
On my local system I've been testing it on latest `main`, but apparently it's testing against `v1.5.8` in this repo. But I'm not sure if that should affect the evaluation since it was released just yesterday.

**Comment by zeborg** [comment]
Update:

I've just verified it's working for kyverno versions including and after this commit: https://github.com/kyverno/kyverno/commit/bb0690111922554ecc36644ad6bd2e19b95168ac

The `v1.5.8` release does not seem to have this fix implemented and therefore the build is failing for this policy as of now.

I believe it's better to have `policies:main` be tested against the `kyverno:main` CLI, same as we currently do for CLI tests in `kyverno:main` within the CI job, and if this change is acceptable, I can create a separate issue and implement it in another PR.

**Comment by chipzoller** [comment]
Yes, kyverno/Kyverno:main should be testing against kyverno/policies:main. Any release branch needs to be tested in a similar mapping. I thought we were already doing this.

**Comment by zeborg** [comment]
We're currently doing this for [kyverno/kyverno](https://github.com/kyverno/kyverno), as in, we're testing the PRs made against `kyverno:main` with `policies:main`, but looking at the workflow in this repo it seems like instead of `kyverno:main`, we're fetching the latest kyverno `release` from [here](https://api.github.com/repos/kyverno/kyverno/releases/latest) and using it in the CI job for PR checks made against `policy:main`.

**Review by chipzoller — COMMENTED** [review]
I think we need to add the annotation to the policies here to disable auto-gen. Although this now works in 1.6.0, we probably don't want to bump min version to that for these just yet.

**Review by chipzoller — APPROVED** [review]

**Pull request merged**